### ISSUE TITLE
Can't upgrade to Opam 2.0 on Windows 10/Cygwin64

### ISSUE THREAD

**Issue by jfehrle** [ocaml/opam#3785]
I've been trying various ways to upgrade from 1.2.2 to 2.0 following the instructions at https://opam.ocaml.org/blog/opam-2-0-0/ without success.  They all seem to fail:

1. `sh <(curl -sL https://raw.githubusercontent.com/ocaml/opam/master/shell/install.sh)` says to download the Opam source, then run `make cold`, which dies after a few thousand lines of output with:

```
'/cygdrive/c/opam/opam-2.0.3/bootstrap/ocaml/bin/ocamldep.opt' -modules src/which_program/which_program.boot.ml > boot-depends.txt
'/cygdrive/c/opam/opam-2.0.3/bootstrap/ocaml/bin/ocamldep.opt' -modules src/xdg/xdg.ml > boot-depends.txt
'/cygdrive/c/opam/opam-2.0.3/bootstrap/ocaml/bin/ocamllex.opt' -q src/let-syntax/lexer.mll
'/cygdrive/c/opam/opam-2.0.3/bootstrap/ocaml/bin/ocamlc.opt' -g -w -40 -o boot.exe unix.cma boot.ml
      1 [main] ocamlrun 16520 child_info_fork::abort: address space needed by 'dllunix.so' (0x400000) is already occupied
Error: fork: : Resource temporarily unavailable
```

2. Similarly `opam update; opam install opam-devel` dies while installing jbuilder:

```
#=== ERROR while installing jbuilder.1.0+beta20 ===============================#
# opam-version 1.2.2
# os           cygwin
# command      ./boot.exe -j 4
# path         /home/Jim/.opam/4.06.1/build/jbuilder.1.0+beta20
# compiler     4.06.1
# exit-code    1
# env-file     /home/Jim/.opam/4.06.1/build/jbuilder.1.0+beta20/jbuilder-5052-e0259d.env
# stdout-file  /home/Jim/.opam/4.06.1/build/jbuilder.1.0+beta20/jbuilder-5052-e0259d.out
# stderr-file  /home/Jim/.opam/4.06.1/build/jbuilder.1.0+beta20/jbuilder-5052-e0259d.err
### stderr ###
#       1 [main] ocamlrun 5124 child_info_fork::abort: address space needed by 'dllunix.so' (0x420000) is already occupied
# Error: fork: : Resource temporarily unavailable
```

Rebooting doesn't help.

This issue appears similar to https://github.com/ocaml/opam/issues/3503, but I'm able to compile many, many files with OCaml.

I saw this suggestion in https://github.com/ocaml/opam/issues/2276 for Cygwin32.  Perhaps there an equivalent for Cygwin64?

```
rebase -b 0x7cd20000 ./.opam/4.02.3/lib/ocaml/stublibs/dllunix.so
rebase -b 0x7cdc0000 ./.opam/4.02.3/lib/ocaml/stublibs/dllthreads.so
```

Should I update to the latest cygwin?  Timestamps suggest I installed in on 11/24/17.

```
$ opam config report
# OPAM config report
# opam-version    1.2.2
# self-upgrade    no
# os              cygwin
# external-solver no
# criteria        -removed,-notuptodate,-changed
# jobs            4
# repositories    1 (http)
# pinned          0
# current-switch  4.06.1
# last-update     2019-03-16 07:58
```

**Comment by jfehrle** [comment]
@dra27, I was told you are the Windows wizard.

**Comment by takanuva** [comment]
I'm having the same issue; I've successfully installed opam 2 from the Cygwin installation wizard, but I get the same error when then trying to install dune through opam.

    0 [main] ocamlrun 6809 child_info_fork::abort: address space needed by 'dllunix.so' (0x400000) is already occupied

I've tried using the trick @dra27 suggested on #2276, but rebase tells me that `0x7cd20000` is too small a value to use, and that the minimum allowed is `0x200000000`.

Does anyone know how to solve this issue?

**Comment by jfehrle** [comment]
@takanuva I was able to install opam2 on cygwin32.  Another alternative is to use Linux for Windows, which appears to run Linux-format executables on top of a Windows environment.  LfW doesn't support Linux graphics/GUI programs, but should support programs with text input and output.  But would be much better to get cygwin64 to work as @dra27 suggested in #2276.

**Comment by rjbou** [comment]
Related to this issue raised by @dra27 in the [cygwin mailing list](https://cygwin.com/ml/cygwin/2018-02/msg00080.html).

**Comment by rhlee** [comment]
`rebase` 4.4.4 prevents you rebasing to below `0x200000000`. Compiling `rebase` 4.4.3 from [git://sourceware.org/git/cygwin-apps/rebase.git](url) seems to work. I don't really understand win32/PE format internals and I'm sure there is a very good reason why you can't rebase to below that boundary, so I'm not sure what problems it will cause. But it seems to work for now and I can compile `dune` using opam 2.

**Comment by dra27** [comment]
@rhlee - see [this message](https://sourceware.org/ml/cygwin/2018-02/msg00097.html) in the thread @rjbou referenced.

TL;DR You'll probably be fine, but you _might_ get fork failure messages more frequently than normal.

I do really want to get the various patches in FlexDLL and OCaml dusted off and completed, but the work is not trivial (in particular the OCaml PR) - I doubt I will have time to get it ready in time for the OCaml 4.09 freeze, but hopefully it'll be done in OCaml 4.10 which will be frozen towards the end of this year.

**Comment by dra27** [comment]
OCaml 4.12 finally restored Cygwin64 support - although the official Cygwin package still needs to be re-adopted to be updated from 4.10 (it's on my list...)
